aten/src/ATen/native/MaxUnpooling.cpp

```diff
@@ -0,0 +1,604 @@
+  AT_CHECK(
+      indices.scalar_type() == at::ScalarType::Long,
+      "elements in indices should be type Long");
+  AT_CHECK(
+      self.sizes() == indices.sizes(), "Input shape must match indices shape");
+
+  AT_CHECK(output_size.size() == 2, "Output size must be 2");
+
+  int owidth = output_size[0];
+  int oheight = output_size[1];
+
+  /* get contiguous gradOutput and indices */
+  auto gradOutput = grad_output.contiguous();
+  auto indicesContiguous = indices.contiguous();
+
+  /* resize */
+  grad_input.resize_as_(self);
+  grad_input.zero_();
+
+  if (self.ndimension() == 4) {
+    nbatch = self.size(0);
+    dimw++;
+    dimh++;
+  }
+
+  /* sizes */
+  nslices = self.size(dimh - 1);
+  iheight = self.size(dimh);
+  iwidth = self.size(dimw);
+
+  if (output_size[0] != gradOutput.size(dimh) ||
+      output_size[1] != gradOutput.size(dimw)) {
+    AT_ERROR(
+        "Inconsistent gradOutput size",
+        output_size[0],
+        ", output width= ",
+        output_size[1],
+        ", gradOutput: ",
+        gradOutput.size(dimh),
+        "x",
+        gradOutput.size(dimw));
+  }
+
+  int p;
+  for (p = 0; p < nbatch; p++) {
+    auto inputOffset = p * nslices * iheight * iwidth;
+    auto outputOffset = p * nslices * oheight * owidth;
+    AT_DISPATCH_FLOATING_TYPES(
+        self.type(), "max_unpooling2d_backward_out_cpu_frame", ([&] {
+          max_unpooling2d_backward_out_cpu_frame<scalar_t>(
+              grad_input.data<scalar_t>() + inputOffset,
+              gradOutput.data<scalar_t>() + outputOffset,
+              indices.data<int64_t>() + inputOffset,
+              nslices,
+              iwidth,
+              iheight,
+              output_size[0],
+              output_size[1]);
+        }));
+  }
+  return grad_input;
+}
+
+Tensor max_unpooling2d_backward_cpu(
+    const Tensor& grad_output,
+    const Tensor& self,
+    const Tensor& indices,
+    IntList output_size) {
+  auto grad_input = at::empty_like(self);
+  max_unpooling2d_backward_out_cpu(
+      grad_input, grad_output, self, indices, output_size);
+  return grad_input;
+}
+
+template <typename scalar_t>
+static void max_unpooling3d_backward_out_cpu_frame(
+    scalar_t* gradInput_p,
+    scalar_t* gradOutput_p,
+    int64_t* ind_p,
+    int64_t nslices,
+    int64_t iT,
+    int64_t iW,
+    int64_t iH,
+    int64_t oT,
+    int64_t oW,
+    int64_t oH) {
+  int k;
+#pragma omp parallel for private(k)
+  for (k = 0; k < nslices; k++) {
+    scalar_t* gradInput_p_k = gradInput_p + k * iT * iH * iW;
+    scalar_t* gradOutput_p_k = gradOutput_p + k * oT * oH * oW;
+    int64_t* ind_p_k = ind_p + k * iT * iH * iW;
+
+    int t, i, j, index;
+    int64_t maxp;
+    for (t = 0; t < iT; t++) {
+      for (i = 0; i < iH; i++) {
+        for (j = 0; j < iW; j++) {
+          index = t * iH * iW + i * iW + j;
+          maxp = ind_p_k[index]; /* retrieve position of max */
+          if (maxp < 0 || maxp >= oT * oH * oW) {
+            AT_ERROR(
```

Comment:
I actually haven't looked into this -- does this work inside the omp parallel?  What's the rule?

Reply:
I used the same `bool has_error` pattern that was used elsewhere as a replacement.